test/integration/puppetlabs/services/master/environment_classes_int_test.clj

```diff
@@ -522,3 +550,123 @@
+          "X_ssl-client-verify-header-FOO"
+          :puppet-version
+          "1.2.3"}))
+
+(deftest ^:integration class-info-updated-after-cache-flush-during-prior-request
+  (let [puppet-server-settings (default-puppet-server-settings)
+        class-info-atom (atom [{"name" "someclass" "params" []}])
+        wait-atom (atom nil)]
+    ;; This test uses a mock jruby instance function which can provide mock
+    ;; data for an environment class info query and can suspend a request
+    ;; long enough for the cached environment data to be invalidated.
+    (with-redefs
+     [jruby-internal/create-pool-instance!
+      (partial jruby-testutils/create-mock-pool-instance
+               (partial create-jruby-instance-with-mock-class-info
+                        wait-atom
+                        class-info-atom))]
+      (bootstrap/with-puppetserver-running-with-services
+       app
+       [handler/request-handler-service
+        jruby-service/jruby-puppet-pooled-service
+        profiler/puppet-profiler-service
+        webserver/jetty9-service
+        webrouting/webrouting-service
+        mock-puppetserver-config-service
+        master-service/master-service
+        ca/certificate-authority-service
+        authorization/authorization-service
+        admin/puppet-admin-service
+        vcs/versioned-code-service]
+       {:jruby-puppet {:max-active-instances 1}
+        :webserver {:ssl-ca-cert (:localcacert puppet-server-settings)
+                    :ssl-cert (:hostcert puppet-server-settings)
+                    :ssl-key (:hostprivkey puppet-server-settings)}
+        :puppet-server puppet-server-settings}
+       (let [continue-promise (promise)
+             wait-promise (promise)
+             _ (reset! wait-atom {:continue-promise continue-promise
+                                  :wait-promise wait-promise})
+             initial-response-future (future (get-env-classes "production"))]
+         @wait-promise
+         (reset! class-info-atom [{"name" "updatedclass"
+                                   "params" []}])
+         (purge-env-cache "production")
+         (deliver continue-promise true)
+         (let [initial-response @initial-response-future
+               initial-response-etag (response-etag initial-response)
+               _ (reset! wait-atom nil)
+               response-after-update (get-env-classes "production"
+                                                      initial-response-etag)]
+           (testing (str "initial request in progress while environment "
+                         "cache is invalidated contains original class info")
+             (is (= 200 (:status initial-response))
+                 (str
+                  "unexpected status code for initial response"
+                  "response: "
+                  (ks/pprint-to-string initial-response)))
+             (is (not (nil? initial-response))
+                 "no etag returned for initial response")
```

Comment:
Should this be asserting that `initial-response-etag` is not nil?  As opposed to `initial-response`?